flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/AbstractAsyncStateStreamOperator.java

```diff
@@ -368,10 +385,12 @@ protected void processWatermarkStatus(WatermarkStatus watermarkStatus, int index
                     boolean wasIdle = combinedWatermark.isIdle();
                     // index is 0-based
                     if (combinedWatermark.updateStatus(index, watermarkStatus.isIdle())) {
-                        super.processWatermark(
-                                new Watermark(combinedWatermark.getCombinedWatermark()));
-                    }
-                    if (wasIdle != combinedWatermark.isIdle()) {
+                        doProcessWatermark(
```

Comment:
This part is basically a rewrite of the overriden method, `AbstractStreamOperator#emitWatermarkStatus`, considering the async timer processing and emitting watermark or status after that. The logic in the method of parent class is more readable. Which is:

-  If a combined watermark changes (because some of the watermarks become active), we process the combined watermark.
- If the combined watermark idle/active status switched, notify the downstream.

These two if(s) are independent.

This PR ensures watermark status emitting after the watermark processing under the async context. I don't think a comment or two would explain this logic well, but you can get the point if you compare this method with the one from parent class.

Reply:
Thanks @Zakelly 
you say the 2 `if`s are independent,  but the second `if `is in an `else`. Shouldn't we remove the `else` so the `if`s are independent?

Reply:
@davidradl `if`s are independent, but the execution order is serial. This PR must ensure in async context, the possible `processWatermark` and `output.emitWatermarkStatus` still executed in order. So if you read the code:
```
if (conditionA) {
  asyncDo(a; then if (conditionB) then b)
} else if (conditionB) {
  do(b)
}
```

Reply:
Ah I see; and if we removed the else effectively we would be doing the synchronous processing incorrectly for the async case. Thanks for the clarification